### PR TITLE
Verify proxies in contract release verification

### LINKED ISSUE #5483: verify-release should verify the storage root of a new proxy
To verify that a contract deploy that will go through governance is safe, verify-release should do the following:

- [x] Verify that the owner of the proxy is governance (i.e. no more changes can be made to the proxy functionality)
- [x] Verify that the owner of the implementation contract in proxy storage is governance (i.e. no more changes can be made to proxy storage wrt. the implementation that is governance)
- [x] Verify that the proxy address has proxy bytecode
- [x] Verify that the all storage variables are unset from the implementation contract in proxy storage (i.e. the deployer did not set any values)

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

- Adds storage verification to `verify-release` when proxy contracts are proposed to be changed 

### Other changes

- Bumps protocol package web3 version from 1.2.4 to 1.2.6 for access to `eth_getProof`
  - See https://github.com/ethereum/EIPs/issues/1186

### Tested

Screen Shot 2020-11-11 at 5 44 02 PM

### Related issues

- Fixes #5483

### Backwards compatibility

Yes.

### CODE REVIEW

File: packages/protocol/lib/proxy-utils.ts [hunk 1]
```diff
@@ -1,8 +1,36 @@
...
+const IMPLEMENTATION_POSITION = '0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc'
+
+export async function verifyProxyStorageProof(web3: Web3, proxy: string, owner: string) {
+  const proof = await web3.eth.getProof(
+    web3.utils.toChecksumAddress(proxy),
+    [OWNER_POSITION, IMPLEMENTATION_POSITION],
+    'latest'
+  )
+
+  const trie = new SecureTrie()
```
Comment (nambrot) [inline packages/protocol/lib/proxy-utils.ts:27]: A little less exciting than I was hoping for :P

**Review by nambrot — APPROVED** [review]

File: packages/protocol/lib/compatibility/verify-bytecode.ts [hunk 2]
```diff
@@ -78,60 +81,63 @@ const getOnchainBytecode = async (address: string, context: VerificationContext)
...
 const dfsStep = async (queue: string[], visited: Set<string>, context: VerificationContext) => {
   const contract = queue.pop()
   // mark current DFS node as visited
   visited.add(contract)
 
-  // get source code
+  // check proxy deployment
+  if (isProxyChanged(contract, context)) {
+    const proxyAddress = getProposedProxyAddress(contract, context)
+    const governanceAddr = await context.registry.getAddressForString('Governance')
```
Comment (m-chrzan) [inline packages/protocol/lib/compatibility/verify-bytecode.ts]: Optional, slight optimalization: we could get the Governance address once and store it in context.

File: packages/protocol/lib/compatibility/verify-bytecode.ts [hunk 3]
```diff
@@ -78,60 +81,63 @@ const getOnchainBytecode = async (address: string, context: VerificationContext)
...
+    }
+
+    const onchainProxyBytecode = await getOnchainBytecode(proxyAddress, context)
+    const sourceProxyBytecode = getSourceBytecode(`${contract}Proxy`, context)
+    if (onchainProxyBytecode !== sourceProxyBytecode) {
+      throw new Error(`Proposed ${contract}Proxy does not match compiled proxy bytecode`)
+    }
+
+    console.log(
+      `Proxy deployed at ${proxyAddress} matches ${contract}Proxy (implementation and storage)`
```
Comment (m-chrzan) [inline packages/protocol/lib/compatibility/verify-bytecode.ts]: Wording nit: I'd change "implementation and storage" to "bytecode and storage", implementation in my mind being reserved for the logic contract pointed to by the Proxy.

File: packages/protocol/lib/proxy-utils.ts [hunk 4]
```diff
@@ -1,8 +1,36 @@
...
+//   uint256(keccak256("eip1967.proxy.admin")) - 1
+// );
+const OWNER_POSITION = '0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103'
+
+// bytes32 private constant IMPLEMENTATION_POSITION = bytes32(
+//   uint256(keccak256("eip1967.proxy.implementation")) - 1
+// );
+const IMPLEMENTATION_POSITION = '0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc'
+
+export async function verifyProxyStorageProof(web3: Web3, proxy: string, owner: string) {
```
Comment (m-chrzan) [inline packages/protocol/lib/proxy-utils.ts:20]: I would love to see some unit tests of this function, both a positive and a negative case.

Comment (m-chrzan) [reply packages/protocol/lib/proxy-utils.ts:20]: Oh, but that won't be possible with ganache... Only option would be to mock web3?

File: packages/protocol/lib/compatibility/verify-bytecode.ts (hunk 2 above)
Comment (yorhodes) [reply packages/protocol/lib/compatibility/verify-bytecode.ts]: now that I think of it, the governance address could change but I will make this optimization for now

**Pull request merged**